internal/core/ebook.go

```diff
@@ -0,0 +1,252 @@
+package core
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	fp "path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/go-shiori/shiori/internal/model"
+	"github.com/pkg/errors"
+)
+
+func EbookGenerate(req ProcessRequest) (book model.Bookmark, isFatalErr bool, err error) {
+	// variable for store generated html code
+	var html string
+
+	book = req.Bookmark
+
+	// Make sure bookmark ID is defined
+	if book.ID == 0 {
+		return book, true, errors.Wrap(err, "bookmark ID is not valid")
+	}
+
+	// cheak archive and thumb
+	strID := strconv.Itoa(book.ID)
+
+	imagePath := fp.Join(req.DataDir, "thumb", fmt.Sprintf("%d", book.ID))
+	archivePath := fp.Join(req.DataDir, "archive", fmt.Sprintf("%d", book.ID))
+
+	if _, err := os.Stat(imagePath); err == nil {
+		book.ImageURL = fp.Join("/", "bookmark", strID, "thumb")
+	}
+
+	if _, err := os.Stat(archivePath); err == nil {
+		book.HasArchive = true
+	}
+	ebookPath := fp.Join(req.DataDir, "ebook", fmt.Sprintf("%d.epub", book.ID))
+	// if epub exist finish prosess else continue
+	if _, err := os.Stat(ebookPath); err == nil {
+		book.HasEbook = true
+		return book, false, nil
+	}
+	contentType := req.ContentType
+	if strings.Contains(contentType, "application/pdf") {
+		return book, true, errors.Wrap(err, "can't create ebook for pdf")
+	}
+
+	ebookDir := fp.Join(req.DataDir, "ebook")
+	// check if directory not exsist create that
+	if _, err := os.Stat(ebookDir); os.IsNotExist(err) {
+		err := os.MkdirAll(ebookDir, model.DataDirPerm)
+		if err != nil {
+			return book, true, errors.Wrap(err, "can't create ebook directory")
+		}
+	}
+	// create epub file
+	epubFile, err := os.Create(ebookPath)
+	if err != nil {
+		return book, true, errors.Wrap(err, "can't create ebook")
+	}
+	defer epubFile.Close()
+
+	// Create zip archive
+	epubWriter := zip.NewWriter(epubFile)
+	defer epubWriter.Close()
+
+	// Create the mimetype file
+	mimetypeWriter, err := epubWriter.Create("mimetype")
+	if err != nil {
+		return book, true, errors.Wrap(err, "can't create mimetype")
+	}
+	_, err = mimetypeWriter.Write([]byte("application/epub+zip"))
+	if err != nil {
+		return book, true, errors.Wrap(err, "can't write into mimetype file")
+	}
+
+	// Create the container.xml file
+	containerWriter, err := epubWriter.Create("META-INF/container.xml")
+	if err != nil {
+		return book, true, errors.Wrap(err, "can't create container.xml")
+	}
+
+	_, err = containerWriter.Write([]byte(`<?xml version="1.0" encoding="UTF-8"?>
```

Comment:
Create issue for: Move ebook content to constant files or assets using embed.

Reply:
I meant moving the XML content to constants somewhere else (at the top of the file?) instead of having them inline in the code's logic.